File: references/betterlectio/ios/BetterLectio/SupabaseSubjectService.swift
//
//  SupabaseSubjectService.swift
//  BetterLectio
//

import Foundation
import Supabase

final class SupabaseSubjectService {
    static let shared = SupabaseSubjectService()

    private let manager: SupabaseManager

    init(manager: SupabaseManager = .shared) {
        self.manager = manager
    }

    func fetchMappings(studentId: String, schoolId: String) async throws -> [SupabaseSubjectMapping] {
        guard let client = manager.client else { return [] }

        return try await client.rpc(
            "get_student_lesson_mappings_v2",
            params: FetchMappingsParams(p_school_id: schoolId, p_student_id: studentId)
        )
        .execute()
        .value
    }

    func upsertMappingOverride(
        studentId: String,
        schoolId: String,
        mappingId: UUID,
        displayName: String?,
        colorHue: Int?,
        icon: String?
    ) async throws {
        guard let client = manager.client else { return }

        let params = UpsertOverrideParams(
            p_school_id: schoolId,
            p_student_id: studentId,
            p_mapping_id: mappingId.uuidString.lowercased(),
            p_display_name: displayName,
            p_color_hue: colorHue,
            p_icon: icon,
            p_client_updated_at: Self.clientTimestamp(),
            p_last_modified_by: "ios"
        )

        try await client.rpc(
            "upsert_user_lesson_override_v2",
            params: params
        )
        .execute()
    }

    func resetMappingOverride(
        studentId: String,
        schoolId: String,
        mappingId: UUID
    ) async throws {
        guard let client = manager.client else { return }

        let params = ResetOverrideParams(
            p_school_id: schoolId,
            p_student_id: studentId,
            p_mapping_id: mappingId.uuidString.lowercased(),
            p_client_updated_at: Self.clientTimestamp(),
            p_last_modified_by: "ios"
        )

        try await client.rpc(
            "reset_user_lesson_override_v2",
            params: params
        )
        .execute()
    }

    private static func clientTimestamp() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: Date())
    }
}

private nonisolated struct FetchMappingsParams: Encodable, Sendable {
    let p_school_id: String
    let p_student_id: String
}

private nonisolated struct UpsertOverrideParams: Encodable, Sendable {
    let p_school_id: String
    let p_student_id: String
    let p_mapping_id: String
    let p_display_name: String?
    let p_color_hue: Int?
    let p_icon: String?
    let p_client_updated_at: String
    let p_last_modified_by: String

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(p_school_id, forKey: .p_school_id)
        try container.encode(p_student_id, forKey: .p_student_id)
        try container.encode(p_mapping_id, forKey: .p_mapping_id)
        try container.encode(p_display_name, forKey: .p_display_name)
        try container.encode(p_color_hue, forKey: .p_color_hue)
        try container.encode(p_icon, forKey: .p_icon)
        try container.encode(p_client_updated_at, forKey: .p_client_updated_at)
        try container.encode(p_last_modified_by, forKey: .p_last_modified_by)
    }

    private enum CodingKeys: String, CodingKey {
        case p_school_id, p_student_id, p_mapping_id
        case p_display_name, p_color_hue, p_icon
        case p_client_updated_at, p_last_modified_by
    }
}

private nonisolated struct ResetOverrideParams: Encodable, Sendable {
    let p_school_id: String
    let p_student_id: String
    let p_mapping_id: String
    let p_client_updated_at: String
    let p_last_modified_by: String
}

struct SupabaseSubjectMapping: Decodable {
    let mappingId: UUID
    let canonicalKey: String
    let defaultName: String
    let defaultColorHue: Int
    let icon: String?
    let displayName: String
    let displayColorHue: Int
    let displayIcon: String?
    let deletedAt: String?

    private enum CodingKeys: String, CodingKey {
        case id
        case mapping_id
        case canonical_key
        case default_name
        case default_color_hue
        case icon
        case default_icon
        case display_name
        case display_color_hue
        case display_icon
        case deleted_at
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        if let mappingId = try container.decodeIfPresent(UUID.self, forKey: .mapping_id) {
            self.mappingId = mappingId
        } else {
            self.mappingId = try container.decode(UUID.self, forKey: .id)
        }

        canonicalKey = try container.decode(String.self, forKey: .canonical_key)
        defaultName = try container.decode(String.self, forKey: .default_name)
        defaultColorHue = try container.decode(Int.self, forKey: .default_color_hue)

        if let icon = try container.decodeIfPresent(String.self, forKey: .icon) {
            self.icon = icon
        } else {
            self.icon = try container.decodeIfPresent(String.self, forKey: .default_icon)
        }

        displayName = try container.decodeIfPresent(String.self, forKey: .display_name) ?? defaultName
        displayColorHue = try container.decodeIfPresent(Int.self, forKey: .display_color_hue) ?? defaultColorHue
        displayIcon = try container.decodeIfPresent(String.self, forKey: .display_icon) ?? icon
        deletedAt = try container.decodeIfPresent(String.self, forKey: .deleted_at)
    }
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/supabase/SupabaseSubjectService.kt
package dk.betterlectio.android.feature.supabase

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import timber.log.Timber
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Singleton

data class SupabaseSubjectMapping(
    val mappingId: String,
    val canonicalKey: String,
    val defaultName: String,
    val defaultColorHue: Int,
    val icon: String?,
    val displayName: String,
    val displayColorHue: Int,
    val displayIcon: String?,
    val deletedAt: String?,
)

/**
 * Subject / lesson mapping sync (iOS: `SupabaseSubjectService`).
 * RPCs: `get_student_lesson_mappings_v2`, `upsert_user_lesson_override_v2`,
 * `reset_user_lesson_override_v2`.
 */
@Singleton
class SupabaseSubjectService @Inject constructor(
    private val manager: SupabaseManager,
) {
    suspend fun fetchMappings(studentId: String, schoolId: String): List<SupabaseSubjectMapping> {
        val client = manager.client ?: return emptyList()
        if (manager.awaitSessionReady() !is SupabaseSessionState.Ready) return emptyList()
        return try {
            client.postgrest.rpc(
                function = "get_student_lesson_mappings_v2",
                parameters = FetchMappingsParams(
                    pSchoolId = schoolId,
                    pStudentId = studentId,
                ),
            ).decodeList<MappingRow>().mapNotNull { it.toMapping() }
        } catch (e: Exception) {
            Timber.w(e, "subject fetchMappings failed")
            emptyList()
        }
    }

    suspend fun upsertMappingOverride(
        studentId: String,
        schoolId: String,
        mappingId: String,
        displayName: String?,
        colorHue: Int?,
        icon: String?,
    ) {
        val client = manager.client ?: return
        if (manager.awaitSessionReady() !is SupabaseSessionState.Ready) return
        try {
            client.postgrest.rpc(
                function = "upsert_user_lesson_override_v2",
                parameters = UpsertOverrideParams(
                    pSchoolId = schoolId,
                    pStudentId = studentId,
                    pMappingId = mappingId.lowercase(),
                    pDisplayName = displayName,
                    pColorHue = colorHue,
                    pIcon = icon,
                    pClientUpdatedAt = clientTimestamp(),
                    pLastModifiedBy = "android",
                ),
            )
        } catch (e: Exception) {
            Timber.w(e, "subject upsertMappingOverride failed")
        }
    }

    suspend fun resetMappingOverride(
        studentId: String,
        schoolId: String,
        mappingId: String,
    ) {
        val client = manager.client ?: return
        if (manager.awaitSessionReady() !is SupabaseSessionState.Ready) return
        try {
            client.postgrest.rpc(
                function = "reset_user_lesson_override_v2",
                parameters = ResetOverrideParams(
                    pSchoolId = schoolId,
                    pStudentId = studentId,
                    pMappingId = mappingId.lowercase(),
                    pClientUpdatedAt = clientTimestamp(),
                    pLastModifiedBy = "android",
                ),
            )
        } catch (e: Exception) {
            Timber.w(e, "subject resetMappingOverride failed")
        }
    }

    @Serializable
    private data class FetchMappingsParams(
        @SerialName("p_school_id") val pSchoolId: String,
        @SerialName("p_student_id") val pStudentId: String,
    )

    @Serializable
    private data class UpsertOverrideParams(
        @SerialName("p_school_id") val pSchoolId: String,
        @SerialName("p_student_id") val pStudentId: String,
        @SerialName("p_mapping_id") val pMappingId: String,
        @SerialName("p_display_name") val pDisplayName: String?,
        @SerialName("p_color_hue") val pColorHue: Int?,
        @SerialName("p_icon") val pIcon: String?,
        @SerialName("p_client_updated_at") val pClientUpdatedAt: String,
        @SerialName("p_last_modified_by") val pLastModifiedBy: String,
    )

    @Serializable
    private data class ResetOverrideParams(
        @SerialName("p_school_id") val pSchoolId: String,
        @SerialName("p_student_id") val pStudentId: String,
        @SerialName("p_mapping_id") val pMappingId: String,
        @SerialName("p_client_updated_at") val pClientUpdatedAt: String,
        @SerialName("p_last_modified_by") val pLastModifiedBy: String,
    )

    @Serializable
    private data class MappingRow(
        val id: String? = null,
        @SerialName("mapping_id") val mappingId: String? = null,
        @SerialName("canonical_key") val canonicalKey: String? = null,
        @SerialName("default_name") val defaultName: String? = null,
        @SerialName("default_color_hue") val defaultColorHue: Int? = null,
        val icon: String? = null,
        @SerialName("default_icon") val defaultIcon: String? = null,
        @SerialName("display_name") val displayName: String? = null,
        @SerialName("display_color_hue") val displayColorHue: Int? = null,
        @SerialName("display_icon") val displayIcon: String? = null,
        @SerialName("deleted_at") val deletedAt: String? = null,
    ) {
        fun toMapping(): SupabaseSubjectMapping? {
            val mid = mappingId ?: id ?: return null
            val key = canonicalKey?.takeIf { it.isNotBlank() } ?: return null
            val defName = defaultName?.takeIf { it.isNotBlank() } ?: key
            val defHue = defaultColorHue ?: 210
            val ico = icon ?: defaultIcon
            return SupabaseSubjectMapping(
                mappingId = mid,
                canonicalKey = key,
                defaultName = defName,
                defaultColorHue = defHue,
                icon = ico,
                displayName = displayName?.takeIf { it.isNotBlank() } ?: defName,
                displayColorHue = displayColorHue ?: defHue,
                displayIcon = displayIcon ?: ico,
                deletedAt = deletedAt,
            )
        }
    }

    companion object {
        private val TIMESTAMP_FORMAT: DateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)

        private fun clientTimestamp(): String = TIMESTAMP_FORMAT.format(Instant.now())

        /** Convert 0–360 hue to ARGB (fixed S/V, matching iOS lesson colors). */
        fun hueToArgb(hue: Int): Long {
            val h = ((hue % 360) + 360) % 360 / 360f
            val s = 0.62f
            val v = 0.88f
            val i = (h * 6).toInt()
            val f = h * 6 - i
            val p = v * (1 - s)
            val q = v * (1 - f * s)
            val t = v * (1 - (1 - f) * s)
            val (r, g, b) = when (i % 6) {
                0 -> Triple(v, t, p)
                1 -> Triple(q, v, p)
                2 -> Triple(p, v, t)
                3 -> Triple(p, q, v)
                4 -> Triple(t, p, v)
                else -> Triple(v, p, q)
            }
            val ri = (r * 255).toInt().coerceIn(0, 255)
            val gi = (g * 255).toInt().coerceIn(0, 255)
            val bi = (b * 255).toInt().coerceIn(0, 255)
            return 0xFF000000L or (ri.toLong() shl 16) or (gi.toLong() shl 8) or bi.toLong()
        }

        fun argbToHue(argb: Long): Int {
            val r = ((argb shr 16) and 0xFF) / 255f
            val g = ((argb shr 8) and 0xFF) / 255f
            val b = (argb and 0xFF) / 255f
            val max = maxOf(r, g, b)
            val min = minOf(r, g, b)
            val d = max - min
            if (d < 1e-6f) return 0
            val h = when (max) {
                r -> ((g - b) / d + if (g < b) 6 else 0)
                g -> (b - r) / d + 2
                else -> (r - g) / d + 4
            } / 6f
            return ((h * 360f).toInt() % 360 + 360) % 360
        }
    }
}
